airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/state/CursorStateMessageProducer.java

```diff
@@ -0,0 +1,144 @@
+        if (!Objects.equals(currentMaxCursor, initialCursor)) {
+          // Only create an intermediate state when it is not the first record.
+          intermediateStateMessage = createStateMessage(stream, totalRecordCount);
+        }
+        currentMaxCursor = cursorCandidate;
+        currentMaxCursorRecordCount = 1L;
+      } else if (cursorComparison == 0) {
+        currentMaxCursorRecordCount++;
+      } else if (cursorComparison > 0) {
+        cursorOutOrOrderDetected = true;
+      }
+    }
+    return message;
+
+  }
+
+  @Override
+  public AirbyteStateMessage createFinalStateMessage(final ConfiguredAirbyteStream stream) {
+    return createStateMessage(stream, totalRecordCount);
+  }
+
+  /**
+   * Only sends out state message when there is a state message to be sent out.
+   */
+  @Override
+  public boolean shouldEmitStateMessage(final ConfiguredAirbyteStream stream) {
+    return intermediateStateMessage != null;
+  }
+
+  /**
+   * Creates AirbyteStateMessage while updating the cursor used to checkpoint the state of records
+   * read up so far
+   *
+   * @param recordCount count of total read messages. Used to determine log frequency.
+   * @return AirbyteMessage which includes information on state of records read so far
+   */
+  private AirbyteStateMessage createStateMessage(final ConfiguredAirbyteStream stream, final int recordCount) {
+    final AirbyteStreamNameNamespacePair pair = new AirbyteStreamNameNamespacePair(stream.getStream().getName(), stream.getStream().getNamespace());
+    final AirbyteStateMessage stateMessage = stateManager.updateAndEmit(pair, currentMaxCursor, currentMaxCursorRecordCount);
+    final Optional<CursorInfo> cursorInfo = stateManager.getCursorInfo(pair);
+
+    // logging once every 100 messages to reduce log verbosity
+    if (recordCount % 100 == 0) {
+      LOGGER.info("State report for stream {} - original: {} = {} (count {}) -> latest: {} = {} (count {})",
+          pair,
+          cursorInfo.map(CursorInfo::getOriginalCursorField).orElse(null),
+          cursorInfo.map(CursorInfo::getOriginalCursor).orElse(null),
+          cursorInfo.map(CursorInfo::getOriginalCursorRecordCount).orElse(null),
+          cursorInfo.map(CursorInfo::getCursorField).orElse(null),
+          cursorInfo.map(CursorInfo::getCursor).orElse(null),
+          cursorInfo.map(CursorInfo::getCursorRecordCount).orElse(null));
+    }
+
+    return stateMessage;
+  }
+
+  private String getCursorCandidate(final String cursorField, AirbyteMessage message) {
+    final String cursorCandidate = message.getRecord().getData().get(cursorField).asText();
+    return (cursorCandidate != null ? replaceNull(cursorCandidate) : null);
+  }
+
+  private String replaceNull(final String cursorCandidate) {
+    if (cursorCandidate.contains("\u0000")) {
+      return cursorCandidate.replaceAll("\u0000", "");
+    }
+    return cursorCandidate;
+  }
```

Comment:
```suggestion
    return (cursorCandidate != null ? cursorCandidate.replaceAll("\u0000", "") : null);
  }
```

Also what does this mean? I see this was taken from existing code. What's wrong with `\u0000`? I'd like to know, and so would anyone else reading this code.

Reply:
I do not know the history here - this was copied from https://github.com/airbytehq/airbyte/blame/master/airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/source/relationaldb/StateDecoratingIterator.java#L97

looks like it tries to solve high cpu usage: https://github.com/airbytehq/airbyte/pull/21523

@subodh1810 or @akashkulk do you know if we still need this?